bindings/rust/bench/src/harness.rs

```diff
@@ -1,163 +1,356 @@
-    pub ec_group: ECGroup,
+    pub kx_group: KXGroup,
     pub sig_type: SigType,
 }
 
 impl CryptoConfig {
-    pub fn new(cipher_suite: CipherSuite, ec_group: ECGroup, sig_type: SigType) -> Self {
+    pub fn new(cipher_suite: CipherSuite, kx_group: KXGroup, sig_type: SigType) -> Self {
         Self {
             cipher_suite,
-            ec_group,
+            kx_group,
             sig_type,
         }
     }
 }
 
-pub trait TlsBenchHarness: Sized {
-    /// Default harness
-    fn default() -> Result<Self, Box<dyn Error>> {
-        Self::new(CryptoConfig::default(), HandshakeType::default(), ConnectedBuffer::default())
+pub trait TlsConnection: Sized {
+    /// Library-specific config struct
+    type Config;
+
+    /// Name of the connection type
+    fn name() -> String;
+
+    /// Default connection (client or server)
+    fn default(mode: Mode) -> Result<Self, Box<dyn Error>> {
+        Self::new(
+            mode,
+            CryptoConfig::default(),
+            HandshakeType::default(),
+            ConnectedBuffer::default(),
+        )
     }
 
+    /// Make a config with given parameters
+    fn make_config(
+        mode: Mode,
+        crypto_config: CryptoConfig,
+        handshake_type: HandshakeType,
+    ) -> Result<Self::Config, Box<dyn Error>>;
+
+    /// Make connection from existing config and buffer
+    fn new_from_config(
+        config: &Self::Config,
+        connected_buffer: ConnectedBuffer,
+    ) -> Result<Self, Box<dyn Error>>;
+
     /// Initialize buffers, configs, and connections (pre-handshake)
     fn new(
+        mode: Mode,
         crypto_config: CryptoConfig,
         handshake_type: HandshakeType,
         buffer: ConnectedBuffer,
-    ) -> Result<Self, Box<dyn Error>>;
+    ) -> Result<Self, Box<dyn Error>> {
+        Self::new_from_config(
+            &Self::make_config(mode, crypto_config, handshake_type)?,
+            buffer,
+        )
+    }
 
-    /// Run handshake on initialized connections
-    /// Returns error if handshake has already completed
+    /// Run one handshake step: receive msgs from other connection, process, and send new msgs
     fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
-    /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
 
-    /// Get negotiated cipher suite
     fn get_negotiated_cipher_suite(&self) -> CipherSuite;
 
-    /// Get whether or negotiated version is TLS1.3
     fn negotiated_tls13(&self) -> bool;
 
-    /// Send application data from connection in harness pair
-    fn send(&mut self, sender: Mode, data: &[u8]) -> Result<(), Box<dyn Error>>;
+    /// Send application data to ConnectedBuffer
+    fn send(&mut self, data: &[u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Read application data from ConnectedBuffer
+    fn recv(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Shrink buffers owned by the connection
+    fn shrink_connection_buffers(&mut self);
```

Comment:
Why do you need this feature? Are you sure that all TLS implementations are doing the same thing when they're shrinking the buffers? I worry they are all doing different things and the comparison won't be meaningful.

Reply:
They each do slightly different things, but the goal is to shrink the buffers as much as a user reasonably could, and to test the memory usage that a user would typically see in a connection. This feature is to enable different memory benching things to play around with.

Reply:
Are you actually using them in the benchmarking tests? I don't see these functions being called.

Reply:
I've been using them in my personal memory benching experiments. I'm going to clean up those experiments and commit then in a PR after this one.

Reply:
In general, I prefer that PRs only include stuff that is needed in the current code, rather than adding stuff for a future PR. It makes it more challenging to review since I can't tell what this extra stuff will be used for. But it's not something I would block this PR on so I'll resolve this comment.